internal/lifecycle/lifecycle.go

```diff
@@ -0,0 +1,108 @@
+// Package lifecycle provides a streamlined way to manage the application
+// lifecycle, queueing up runs and waiting for them to finish, returning all
+// errors that occurred during the runs.
+package lifecycle
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ErrTimeoutReached is returned if the lifecycle times out.
+var ErrTimeoutReached = errors.New("timeout reached")
+
+// Lifecycle manages the application lifecycle.
+type Lifecycle struct {
+	timeout time.Duration
+	counter sync.WaitGroup
+
+	firstRun sync.Once
+	quitMu   sync.RWMutex
+	appMu    sync.Mutex
+
+	errors []error
+
+	exitRequested bool
+	quitted       chan struct{}
+	started       chan struct{}
+}
+
+// New returns a new Lifecycle with the given timeout.
+func New(timeout time.Duration) *Lifecycle {
+	return &Lifecycle{
+		timeout: timeout,
+		quitted: make(chan struct{}),
+		started: make(chan struct{}),
+	}
+}
+
+// Start starts or extends the lifecycle one run and blocks the next run until
+// the current run finishes.
+func (l *Lifecycle) Start() {
```

Comment:
I feel Start() to be a little bit ambiguous. as it’s not the Start of the Lifecycle object, but that we have a new run going. I think we should have something like AddRun() and DoneRun() (you have RunDone() which was already similar)